### ISSUE TITLE
False positive with `unbound variable` under specific conditions

### ISSUE THREAD

**Issue by kdowolski** [microsoft/pyright#6704]
I found that under very specific conditions `pyright` wrongly inform about `unbound variable`.

```python
def foo():
    foo2()

def foo2():
    a = A()
    try:
        with open("path"):
            return
    finally:
        a.foo3() # error: "a" is unbound (reportUnboundVariable)

class A:
    def foo3(self):
        pass
```

- We have some function with `try: finally` block
- In `try` there is some a `contextmanager` that returns something
- There is another function that calls the first one
- The funniest part - that second function must be declared before the first one

**Comment by erictraut** [comment]
Thanks for the bug report. I'm surprised no one has reported this problem previously.

The `finally` clause presents a challenging problem for static analyzers because the code flow at the bottom of the finally clause depends on whether it was entered with an exception or non-exception condition. If it's an exception condition, the exception is re-raised at the bottom of the finally clause. This means it can't be represented as a simple edge in the code flow graph. Pyright has a bunch of logic to deal with this complexity. The bug that caused this incorrect behavior was within this logic.

This will be addressed in the next release. I've added a variant of your code sample to the pyright unit tests to help prevent future regressions.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.340, which I just published. It will be included in a future release of pylance.
